crates/bevy_render/src/batching/mod.rs

```diff
@@ -0,0 +1,123 @@
+        BatchMeta {
+            pipeline_id: item.cached_pipeline(),
+            draw_function_id: item.draw_function(),
+            dynamic_offset: item.dynamic_offset(),
+            user_data,
+        }
+    }
+}
+
+/// A trait to support getting data used for batching draw commands via phase
+/// items.
+pub trait GetBatchData {
+    type Query: ReadOnlyWorldQuery;
+    /// Data used for comparison between phase items. If the pipeline id, draw
+    /// function id, per-instance data buffer dynamic offset and this data
+    /// matches, the draws can be batched.
+    type CompareData: PartialEq;
+    /// The per-instance data to be inserted into the [`GpuArrayBuffer`]
+    /// containing these data for all instances.
+    type BufferData: GpuArrayBufferable + Sync + Send + 'static;
+    /// Get the per-instance data to be inserted into the [`GpuArrayBuffer`].
+    fn get_buffer_data(query_item: &QueryItem<Self::Query>) -> Self::BufferData;
+    /// Get the data used for comparison when deciding whether draws can be
+    /// batched.
+    fn get_compare_data(query_item: &QueryItem<Self::Query>) -> Self::CompareData;
+}
+
+/// Batch the items in a render phase. This means comparing metadata needed to draw each phase item
+/// and trying to combine the draws into a batch.
+pub fn batch_and_prepare_render_phase<I: CachedRenderPipelinePhaseItem, F: GetBatchData>(
+    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
+    mut views: Query<&mut RenderPhase<I>>,
+    query: Query<(Has<NoAutomaticBatching>, F::Query)>,
+) {
+    let gpu_array_buffer = gpu_array_buffer.into_inner();
+
+    let mut process_item = |item: &mut I| {
+        let (no_auto_batching, batch_query_item) = query.get(item.entity()).ok()?;
+
+        let buffer_data = F::get_buffer_data(&batch_query_item);
+        let buffer_index = gpu_array_buffer.push(buffer_data);
+
+        let index = buffer_index.index.get();
+        *item.batch_range_mut() = index..index + 1;
+        *item.dynamic_offset_mut() = buffer_index.dynamic_offset;
+
+        (!no_auto_batching).then(|| {
+            let compare_data = F::get_compare_data(&batch_query_item);
+            BatchMeta::new(item, compare_data)
+        })
+    };
+
+    for mut phase in &mut views {
+        let items = phase.items.iter_mut().map(|item| {
+            let batch_data = process_item(item);
+            (item.batch_range_mut(), batch_data)
+        });
+        items.reduce(|(start_range, prev_batch_meta), (range, batch_meta)| {
+            if batch_meta.is_some() && prev_batch_meta == batch_meta {
+                start_range.end = range.end;
+                (start_range, prev_batch_meta)
+            } else {
+                (range, batch_meta)
+            }
+        });
+    }
+}
+
+pub fn write_batched_instance_buffer<F: GetBatchData>(
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
```

Comment:
Not a huge deal, but theres risk here that two GetBatchData impls will step on each other. Ex: pretend two GetBatchData impls both set `BufferData` to `Mat4`. It would be safer to either use a wrapper Resource over GpuArrayBuffer that is generic on `F`, or to tie GpuArrayBuffer to `F` instead of `F::BufferData`.

Reply:
Makes sense. They have been generic on things like MeshUniform so far so it hasn't been a problem. I've been wanting to rename MeshUniform to something like PerInstanceData or something. I feel like I would want to rework it as part of that. What do you think? I can do it as an immediate follow-up.

Reply:
I'm ok with deferring as long as it doesn't get dropped!